Port replication_pad2d and replication_pad3d to ATen

lhuang04 opened the pull request:
Summary:
port replication padding 2D and 3D from legacy TH API implementation
to ATen implementation.

Differential Revision: D13547567

gchanan reviewed (commented):
can you benchmark?

lhuang04 commented on aten/src/ATen/native/ReplicationPadding.cpp:
```diff
@@ -211,12 +211,625 @@ Tensor& replication_pad1d_backward_out_cpu_template(
+  owidth  = iwidth + pad_l + pad_r;
+
+  AT_CHECK(owidth == gradOutput_.size(dimw),
+      "gradOutput width unexpected. Expected: ", owidth, ", Got: ",
```
I tested, there should be no space automatically added after the comma. see the output of "inputtest", it would be "input test" if there is a space.
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/lhuang04/local/miniconda/lib/python3.7/site-packages/torch/nn/functional.py", line 2686, in pad
    ret = torch._C._nn.replication_pad1d(input, pad)
RuntimeError: non-empty 2D or 3D (batch mode) tensor expected for inputtest

 56   AT_CHECK(input.numel() > 0
 57       && (input.ndimension() == 2 || input.ndimension() == 3),
 58       "non-empty 2D or 3D (batch mode) tensor expected for input", "test");

lhuang04 commented:
# after fold in omp call to address the issue of omp outside dispatch (V3)
# fold omp call
>>> timeit("foo(lst)", globals=globals(), number=10)
3.1397787330206484
>>> timeit("foo(clst)", globals=globals(), number=10)
0.026578820077702403

ezyang commented on aten/src/ATen/native/ReplicationPadding.cpp:
```diff
@@ -1,150 +1,143 @@
-#pragma omp parallel for private(k, ip_x)
-  for (k = 0; k < nslices; k++)
+  long p, k, ip_x;
+#pragma omp parallel for private(p, k, ip_x)
```
Meh, there is a downside to doing it this way, which is that you are now calling into an OpenMP parallelized loop even when batch size is 1, which comes at a cost. I don't remember if OpenMP is smart enough not to ship the work to a worker thread when this happens but some quick eyeballing of the assembly suggests the answer is no: https://godbolt.org/z/hAV-gF

So, if you wanted to defend this choice, you would have to show via benchmarks that this doesn't actually matter. But in your posted CPU timeits things DID get slower. So I don't want to allow this pessimization for no good reason.

ezyang replied:
(The alternate way, is to continue to have two codepaths for the batch and non-batch case, and have one extra little stub function for the batch case which sets up the OMP parallelization before calling the non-batch function.)

See also discussion in #15480.

lhuang04 commented:
# refactor back to two code path (V6)
>>> timeit("foo(lst)", globals=globals(), number=10)
3.2285069869831204

>>> timeit("foo(lst)", globals=globals(), number=100)
32.892865078058094

ezyang commented on aten/src/ATen/native/ReplicationPadding.cpp:
```diff
@@ -211,12 +211,625 @@ Tensor& replication_pad1d_backward_out_cpu_template(
+  int oStartX = fmax(0, pad_l);
+  int oStartY = fmax(0, pad_t);
+
+  int64_t k, ip_x, ip_y;
```
Doesn't the OMP pragma throw a hissy fit if the loop variables aren't pre-declared?

ezyang commented on aten/src/ATen/native/ReplicationPadding.cpp:
```diff
@@ -186,41 +216,691 @@ Tensor& replication_pad1d_backward_out_cpu_template(
+  int64_t owidth  = iwidth + pad_l + pad_r;
+
+  AT_CHECK(owidth >= 1 || oheight >= 1,
+      "input (H: ", iheight, ", W: ", iwidth, " )is too small."
```
nit: you probably wanted the space after the closing paren

ezyang commented on aten/src/ATen/native/ReplicationPadding.cpp:
```diff
@@ -186,41 +216,691 @@ Tensor& replication_pad1d_backward_out_cpu_template(
+    const Tensor& input,
+    IntList paddingSize)
+{
+  AT_CHECK(paddingSize.size() == 4, "padding Size is expected to be 4");
```
nit: don't capitalize Size?

ezyang commented on aten/src/ATen/native/ReplicationPadding.cpp:
```diff
@@ -268,5 +948,88 @@ Tensor replication_pad1d_backward_cpu(
+    const Tensor& input,
+    IntList paddingSize)
+{
+  auto gradInput = at::zeros_like(input);
```
IIUC, you're going to zero gradInput inside the function anyway, so no need to zero it here. `empty_like` would be sufficient.

ezyang commented on aten/src/ATen/native/cuda/ReplicationPadding.cu:
```diff
@@ -184,26 +311,406 @@ void replication_pad1d_backward_out_cuda_template(
+        dim3 gridSize(THCCeilDiv(outputPlaneSize, 256),
+            devOutput.size(1),
+            devOutput.size(0));
+        dim3 blockSize(outputPlaneSize > 256 ? 256 : outputPlaneSize);
```
Hmm... in the original, it wasn't necessary to copy paste the block here and below. I guess the problem is that there is no default constructor in PackedTensorAccessor. Not sure if we should add one. I guess this is OK for now.

ezyang replied:
I think I'm going to ask that you use a ternary here for the definition of `devInput` and `devOutput`, so you can avoid having to duplicate the epilogue. Something like: `auto devInput = (numInputDims == 3 ? input.unsqueeze(0) :  input).packed_accessor<scalar_t, 4>()` and likewise for the output

ezyang commented on aten/src/ATen/native/cuda/ReplicationPadding.cu:
```diff
@@ -184,26 +311,406 @@ void replication_pad1d_backward_out_cuda_template(
+    int pleft, int pright,
+    int ptop, int pbottom,
+    int pfront, int pback) {
+  AT_CHECK(at::cuda::detail::canUse32BitIndexMath(input), 
```
I'm not the biggest fan of the straight up copy paste of the prelude of this function here. Why not something like optional<Tensor> gradOutput, so that you can test if a gradOutput was passed in or not?

ezyang commented on aten/src/ATen/native/cuda/ReplicationPadding.cu:
```diff
@@ -184,26 +311,406 @@ void replication_pad1d_backward_out_cuda_template(
+        replication_pad_forward_kernel3d <<<gridSize, blockSize, 0,
+                                       at::cuda::getCurrentCUDAStream()>>>(
+                                           devInput, devOutput, pfront, pback, ptop, pbottom, pleft, pright);
+      }
```
nit: indentation here is funny

ezyang commented on aten/src/ATen/native/cuda/ReplicationPadding.cu:
```diff
@@ -184,26 +311,406 @@ void replication_pad1d_backward_out_cuda_template(
+
+      replication_pad_forward_kernel3d <<<gridSize, blockSize, 0,
+      at::cuda::getCurrentCUDAStream()>>>(
+          devInput, devOutput, pfront, pback, ptop, pbottom, pleft, pright);
```
I did not see an error, but it was pretty disconcerting to see `pfront` and `pback` suddenly show up at the beginning of the argument list, even though they were in the back of paddingSize

ezyang reviewed (approved):
Good work. I don't see any correctness problems; there are some comments would be nice to resolve, but Imma go ahead and approve this.

lhuang04 committed:
Port replication_pad2d and replication_pad3d to ATen (#15538)

Summary:
port replication padding 2D and 3D from legacy TH API implementation
to ATen implementation.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/15538

Differential Revision: D13547567

fbshipit-source-id: 5f7073223652aa2743a4d6f6b8c6b393791b59e5
